### PR TITLE
Fix MPS device detection and deprecated GradScaler API

### LINKED ISSUE #635: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated.
train.py:196: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler(enabled=(dtype == 'float16'))
num decayed parameter tensors: 26, with 10,755,456 parameters
num non-decayed parameter tensors: 13, with 4,992 parameters
using fused AdamW: True
compiling the model... (takes a ~minute)
W0903 20:13:19.043000 55763 torch/_inductor/utils.py:1436] [0/0] Not enough SMs to use max_autotune_gemm mode

### LINKED ISSUE #654: MPS + float16: GradScaler / autocast behavior in train.py
Hi,

I’m 12 years old and was reading the train.py script. I noticed that it sets:

<pre> dtype = 'bfloat16' if torch.cuda.is_available() and torch.cuda.is_bf16_supported() else 'float16' 
scaler = torch.cuda.amp.GradScaler(enabled=(dtype == 'float16'))
ctx = nullcontext() if device_type == 'cpu' else torch.amp.autocast(device_type=device_type, dtype=ptdtype) </pre>

Since MPS isn’t CUDA, I was wondering if fp16/GradScaler would behave correctly there. Should the code explicitly disable fp16 when device_type == 'mps', or is it intended to rely on PyTorch’s default handling?

Thanks for making such an awesome repo!

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
## Summary

- **Fix `device_type` detection to handle MPS (Apple Silicon).** The current code only distinguishes between `'cuda'` and `'cpu'`, so setting `device = 'mps'` causes `device_type` to be `'cpu'`. This silently skips `torch.amp.autocast` (uses `nullcontext()` instead), losing mixed-precision benefits on MPS. The fix adds proper `'mps'` detection in `train.py`, `sample.py`, and `bench.py`.

- **Replace deprecated `torch.cuda.amp.GradScaler` with `torch.amp.GradScaler`.** The old API emits a `FutureWarning` and will be removed in a future PyTorch release. The scaler is now only enabled for CUDA + float16, where loss scaling is actually needed and supported.

Fixes #654, fixes #635.

## Test plan

- [ ] Run `python train.py --device=mps` on an Apple Silicon Mac and verify autocast is active (not `nullcontext`)
- [ ] Run `python train.py --device=cuda` on a CUDA machine and verify GradScaler works with float16
- [ ] Run `python train.py --device=cpu` and verify training still works (scaler disabled, no autocast)
- [ ] Confirm no `FutureWarning` about `torch.cuda.amp.GradScaler` is emitted

### CODE REVIEW

**Comment by Mr-Neutr0n** [comment]
Closing this one to tidy up my open pull requests.

It's been open around five months with no review activity, and I'd rather withdraw it than leave it sitting in your queue. Nothing needed from you, and no hard feelings at all.

If the fix is still wanted, this can be reopened, or I'm happy to redo it properly against current main. Apologies for the noise.
